### DISCUSSION TITLE
How to check if a font exists

### DISCUSSION THREAD

**Discussion by jiaxincao** [wezterm/wezterm#3202] [General]
Is there an API that we can use to check if a font exists or not?

I'm asking because I have a git managed config file which is used for mac, windows and Linux. However, it is hard to find a font that all of the systems have. So I'm thinking I can have a list of them, and use the API to check which one exists, and then load the font.

**Comment by nekowinston** [comment]
I think you could use [`wezterm.font_with_fallback`](https://wezfurlong.org/wezterm/config/lua/wezterm/font_with_fallback.html) for this purpose. If a font isn't on a system, it will use the next one.

**Reply by jiaxincao** [reply]
According to this [page](https://wezfurlong.org/wezterm/config/lua/wezterm/font_with_fallback.html),

> Glyphs are looked up in the first font in the list but if missing the next font is checked and so on.

IIUC, the list of fonts are used to serve glyphs in order, which means if a glyph can be found in the first font, wezterm will use it, otherwise, wezterm will try the next font until it can find the glyph in the font.

However, if a font doesn't exist at all, wezterm will still raise an error. For my case, here's the lua script:

```
local fonts = wezterm.font_with_fallback({"Terminus2 (TTF)", "Monaco"})
```

and this is the error I'm seeing

```
Unable to load a font specified by your font=wezterm.font('Terminus2 (TTF)',
{weight="Regular", stretch='Normal', style=Normal}) configuration. Fallback(s)
are being used instead, and the terminal may not render as intended. See
https://wezfurlong.org/wezterm/config/fonts.html for more information

Unable to load a font specified by your font=wezterm.font('Terminus2 (TTF)',
{weight="Regular", stretch='Normal', style=Normal}) configuration. Fallback(s)
are being used instead, and the terminal may not render as intended. See
https://wezfurlong.org/wezterm/config/fonts.html for more information

Unable to load a font specified by your font=wezterm.font('Terminus2 (TTF)',
{weight="Regular", stretch='Normal', style=Normal}) configuration. Fallback(s)
are being used instead, and the terminal may not render as intended. See
https://wezfurlong.org/wezterm/config/fonts.html for more information
```

**Reply by wez** [reply]
The way I avoid this issue is that I have my preferred fonts checked into my dotfiles repo and I use https://wezfurlong.org/wezterm/config/lua/config/font_dirs.html to point to that location. That way I have the same fonts available to wezterm everywhere.

**Reply by jiaxincao** [reply]
Thanks for letting me know, @wez ! I didn't know we can do that. This is a brilliant idea!
